### PR TITLE
routines for computing mean flow quantities and turbulent kinetic energy

### PR DESCRIPTION
**Pull request by boyceg**

### CODE REVIEW

File: include/ibamr/INSStaggeredHierarchyIntegrator.h [hunk 1]
```diff
@@ -432,6 +432,14 @@ class INSStaggeredHierarchyIntegrator : public INSHierarchyIntegrator
...
      */
     int d_Omega_Norm_idx, d_U_regrid_idx, d_U_src_idx, d_indicator_idx, d_F_div_idx;
+
+    /*
+     * Data for computing mean flow quantities.
+     */
+    unsigned int d_mean_flow_averaging_interval = 0;
+    SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > d_U_mean_var, d_UU_mean_var;
+    int d_U_mean_current_idx, d_U_mean_new_idx, d_U_mean_scratch_idx;
+    int d_UU_mean_current_idx, d_UU_mean_new_idx, d_UU_mean_scratch_idx;
```
Comment (drwells) [inline include/ibamr/INSStaggeredHierarchyIntegrator.h]: Please initialize these with `IBTK::invalid_index`.

File: src/navier_stokes/INSStaggeredHierarchyIntegrator.cpp [hunk 2]
```diff
@@ -925,11 +919,10 @@ INSStaggeredHierarchyIntegrator::initializeHierarchyIntegrator(Pointer<PatchHier
         if (d_output_U)
         {
             d_visit_writer->registerPlotQuantity("U", "VECTOR", d_U_cc_idx, 0, d_U_scale);
-            for (unsigned int d = 0; d < NDIM; ++d)
+            for (unsigned int i = 0; i < NDIM; ++i)
```
Comment (drwells) [inline src/navier_stokes/INSStaggeredHierarchyIntegrator.cpp]: I think we usually use `d` for loops over dimensions so I would rather keep it that way.

Comment (boyceg) [reply src/navier_stokes/INSStaggeredHierarchyIntegrator.cpp]: FWIW, I am thinking here in terms of vector/tensor indices (i,j).

Comment (boyceg) [reply src/navier_stokes/INSStaggeredHierarchyIntegrator.cpp]: I have made this a bit more explicit. Let me know if the current state looks OK.

File: doc/news/changes/major/20190507BoyceGriffith [hunk 3]
```diff
@@ -0,0 +1 @@
+New: Added routines for tracking mean flow quantities and computing turbluent kinetic energy and Reynolds stresses for "steady" fluctuating flows.
```
Comment (drwells) [inline doc/news/changes/major/20190507BoyceGriffith]: Would you please add
```html
<br>
(Boyce Griffith, 2019/05/07)
```
to the end and also wrap the text in line one?

File: include/ibamr/INSStaggeredHierarchyIntegrator.h [hunk 4]
```diff
@@ -411,27 +411,44 @@ class INSStaggeredHierarchyIntegrator : public INSHierarchyIntegrator
...
-    int d_P_current_idx, d_P_new_idx, d_P_scratch_idx;
-    int d_F_current_idx, d_F_new_idx, d_F_scratch_idx;
-    int d_Q_current_idx, d_Q_new_idx, d_Q_scratch_idx;
-    int d_N_old_current_idx, d_N_old_new_idx, d_N_old_scratch_idx;
+    int d_U_current_idx = IBTK::invalid_index, d_U_new_idx = IBTK::invalid_index, d_U_scratch_idx = IBTK::invalid_index;
+    int d_P_current_idx = IBTK::invalid_index, d_P_new_idx = IBTK::invalid_index, d_P_scratch_idx = IBTK::invalid_index;
+    int d_F_current_idx = IBTK::invalid_index, d_F_new_idx = IBTK::invalid_index, d_F_scratch_idx = IBTK::invalid_index;
+    int d_Q_current_idx = IBTK::invalid_index, d_Q_new_idx = IBTK::invalid_index, d_Q_scratch_idx = IBTK::invalid_index;
+    int d_N_old_current_idx = IBTK::invalid_index, d_N_old_new_idx = IBTK::invalid_index,
+        d_N_old_scratch_idx = IBTK::invalid_index;
```
Comment (drwells) [inline include/ibamr/INSStaggeredHierarchyIntegrator.h]: :+1:

**Comment by boyceg** [comment]
@drwells, does this now pass muster?

**Comment by drwells** [comment]
Probably. I'll take a closer look in an hour or so.

**Comment by drwells** [comment]
The first commit looks good. Thanks a lot for taking the time to tidy things up. I am looking at the second commit now.

**Comment by boyceg** [comment]
@xxjaehoxx or I will write a test.

**Pull request merged**